ddtrace/internal/encoding.py

```diff
@@ -48,17 +51,36 @@ def encode(self, obj):
         raise NotImplementedError()
 
 
-class JSONEncoder(_EncoderBase):
+class JSONEncoder(json.JSONEncoder, _EncoderBase):
     content_type = "application/json"
 
+    def __init__(self):
+        # Reduce unnecessary whitespace
+        super(JSONEncoder, self).__init__(separators=(",", ":"), indent=None)
+
     def encode_traces(self, traces):
-        normalized_traces = [[span.to_dict() for span in trace] for trace in traces]
+        normalized_traces = [[JSONEncoder._normalize_span(span.to_dict()) for span in trace] for trace in traces]
         return self.encode(normalized_traces)
 
     @staticmethod
-    def encode(obj):
-        # type: (Any) -> str
-        return json.dumps(obj)
+    def _normalize_span(span):
+        # Ensure all string attributes are actually strings and not bytes
```

Comment:
Should we note here that we should technically do this for meta/metrics but are avoiding it for now due to performance reasons?

Reply:
hmm 🤔 

```suggestion
        # Ensure all string attributes are actually strings and not bytes
        # TODO: We should do this for meta/metrics, and other string properties.
```

?

Reply:
idk if it's a TODO since we're intentionally not doing it but know that it's an issue

Reply:
I was trying to decide between a `TODO` and a `DEV`.

```suggestion
        # Ensure all string attributes are actually strings and not bytes
        # DEV: Skip meta/metrics since we already normalize those keys/values to strings
```

?

```suggestion
        # Ensure all string attributes are actually strings and not bytes
        # DEV: Skip meta/metrics to reduce performance impact of normalizing strings
```

Reply:
I'd still like to say that we know that there's a potential decoding error but this is fine to me as-is too

Reply:
but is there? we already handle coercing values in `set_tag`/`set_metric` ?

Reply:
```python
span.meta["tag"] = b"dfajkfj😕"
```

i suppose

Reply:
ugh... yeah...

Reply:
added a note.